R/repl.R

```diff
@@ -0,0 +1,117 @@
+#' Run a Python REPL
+#' 
+#' This function provides a Python REPL in the R session, which can be used
+#' to interactively run Python code. All code executed within the REPL is
+#' run within the Python 'main' module, and will remain accessible after the
+#' REPL is detached.
+#' 
+#' @import utils
+#' 
+#' @export
+py_repl <- function() {
+ 
+  codeop <- import("codeop", convert = TRUE)
+  
+  # check to see if the current environment supports history
+  has_history <- tryCatch(
+    { utils::savehistory(tempfile()); TRUE },
+    error = function(e) FALSE
+  )
+  
+  if (has_history) {
+    
+    # if we have history, save and then restore the current R history
+    utils::savehistory()
+    on.exit(utils::loadhistory(), add = TRUE)
+    
+    # file to be used for command history during session
+    histfile <- getOption("reticulate.repl.histfile")
+    if (is.null(histfile))
+      histfile <- file.path(tempdir(), ".reticulatehistory")
+    
+    # load history (create emptu file if none exists yet)
+    if (!file.exists(histfile))
+      file.create(histfile)
+    utils::loadhistory(histfile)
+    
+  }
+  
+  # buffer of pending console input (we don't send input to
+  # the console until we have a complete Python statement)
+  buffer <- character()
+  
+  # command compiler (used to check if we've received a complete piece
+  # of Python input)
+  compiler <- codeop$CommandCompiler()
+  
+  # record whether the used has requested a quit
+  quit_requested <- FALSE
+  
+  repl <- function() {
+    
+    # read user input
+    prompt <- if (length(buffer)) "... " else ">>> "
+    contents <- readline(prompt = prompt)
+    
+    # special handling for e.g. 'quit', 'exit'
+    if (contents %in% c("quit", "exit")) {
```

Comment:
do we need to trim whitespace before the comparison?

Reply:
Yes we should!